pkg/controller/pxc/controller_test.go

```diff
@@ -57,3 +69,450 @@ var _ = Describe("PerconaXtraDB Cluster", Ordered, func() {
+				Expect(err).To(Succeed())
+			}
+		})
+
+		When("PXC cluster is deleted with delete-proxysql-pvc finalizer sts and pvc should be removed and secrets kept", func() {
+			It("should delete PXC cluster and reconcile changes", func() {
+				Expect(k8sClient.Delete(ctx, cr)).Should(Succeed())
+
+				_, err := reconciler().Reconcile(ctx, ctrl.Request{NamespacedName: crNamespacedName})
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("controller should remove sts", func() {
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, types.NamespacedName{
+						Name:      cr.Name + "-proxysql",
+						Namespace: cr.Namespace,
+					}, &sfsWithOwner)
+					return k8serrors.IsNotFound(err)
+				}, time.Second*15, time.Millisecond*250).Should(BeTrue())
+
+			})
+
+			It("controller should remove pvc for proxysql", func() {
+				pvcList := corev1.PersistentVolumeClaimList{}
+				Eventually(func() bool {
+					err := k8sClient.List(ctx, &pvcList, &client.ListOptions{
+						Namespace: cr.Namespace,
+						LabelSelector: labels.SelectorFromSet(map[string]string{
+							"app.kubernetes.io/component": "proxysql",
+						}),
+					})
+					return err == nil
+				}, time.Second*15, time.Millisecond*250).Should(BeTrue())
+
+				for _, pvc := range pvcList.Items {
+					By(fmt.Sprintf("checking pvc/%s", pvc.Name))
+					Expect(pvc.DeletionTimestamp).ShouldNot(BeNil())
+				}
+			})
+
+			It("controller should keep secrets", func() {
+				secret := &corev1.Secret{}
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, types.NamespacedName{
+						Namespace: cr.Namespace,
+						Name:      cr.Spec.SecretsName,
+					}, secret)
+
+					return k8serrors.IsNotFound(err)
+				}, time.Second*15, time.Millisecond*250).Should(BeFalse())
+
+				Eventually(func() bool {
+					err := k8sClient.Get(ctx, types.NamespacedName{
+						Namespace: cr.Namespace,
+						Name:      "internal-" + cr.Name,
+					}, secret)
+
+					return k8serrors.IsNotFound(err)
+				}, time.Second*15, time.Millisecond*250).Should(BeFalse())
+
+			})
+		})
+	})
+})
+
+var _ = Describe("Finalizer delete-pxc-pvc", Ordered, func() {
+	ctx := context.Background()
+
+	const crName = "del-pxc-pvc-fnlz"
+	const ns = "del-pxc-pvc-fnlz"
+	crNamespacedName := types.NamespacedName{Name: crName, Namespace: ns}
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ns,
+			Namespace: ns,
+		},
+	}
+
+	BeforeAll(func() {
+		By("Creating the Namespace to perform the tests")
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).To(Not(HaveOccurred()))
+
+		_, err = envtest.InstallCRDs(cfg, envtest.CRDInstallOptions{
+			Paths: []string{filepath.Join("testdata", "cert-manager.yaml")},
+		})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterAll(func() {
+		By("Deleting the Namespace to perform the tests")
+		_ = k8sClient.Delete(ctx, namespace)
+	})
+
+	Context("delete-pxc-pvc finalizer specified", Ordered, func() {
+
+		cr, err := readDefaultCR(crName, ns)
```

Comment:
You have to update cr after you assert("should read default cr.yaml") it is Ok, because if it fails this will panic because cr will be nill if some error is returned.

Reply:
145307b